File: 1erdepartamental/python/4.sobrecarga.py
# ============================================
# PROGRAMA: Sobrecarga de Métodos con POO
# DESCRIPCIÓN: Simula sobrecarga usando parámetros por defecto
# ============================================

from abc import ABC, abstractmethod


class IOperadorFlexible(ABC):
    """Interfaz para operadores que aceptan cantidad variable de argumentos"""

    @abstractmethod
    def operar(self, *args):
        pass


class CalculadoraSobrecarga(IOperadorFlexible):
    """
    Simula sobrecarga de métodos con parámetros por defecto.

    Python no soporta sobrecarga nativa, pero se simula con defaults.
    """

    VALOR_DEFECTO = 0
    MULTIPLICADOR_DEFECTO = 1

    def __init__(self):
        self.__resultado = 0

    @property
    def resultado(self):
        return self.__resultado

    def sumar(self, a=VALOR_DEFECTO, b=VALOR_DEFECTO, c=VALOR_DEFECTO):
        """Suma 1, 2 o 3 números"""
        self.__resultado = a + b + c
        return self.__resultado

    def restar(self, a=VALOR_DEFECTO, b=VALOR_DEFECTO, c=VALOR_DEFECTO):
        """Resta sucesiva: a - b - c"""
        self.__resultado = a - b - c
        return self.__resultado

    def multiplicar(
        self, a=MULTIPLICADOR_DEFECTO, b=MULTIPLICADOR_DEFECTO, c=MULTIPLICADOR_DEFECTO
    ):
        """Multiplica 1, 2 o 3 números"""
        self.__resultado = a * b * c
        return self.__resultado

    def dividir(
        self, a=MULTIPLICADOR_DEFECTO, b=MULTIPLICADOR_DEFECTO, c=MULTIPLICADOR_DEFECTO
    ):
        """División sucesiva con validación"""
        if b != 0 and c != 0:
            self.__resultado = a / b / c
            return self.__resultado
        return None

    def operar(self, *args):
        """Suma cantidad variable de argumentos"""
        return sum(args)

    def __str__(self):
        return f"CalculadoraSobrecarga(resultado={self.__resultado})"


# ============================================
# EJECUCIÓN
# ============================================
if __name__ == "__main__":
    c = CalculadoraSobrecarga()

    # Demuestra "sobrecarga" con diferente cantidad de argumentos
    print(f"sumar() = {c.sumar()}")
    print(f"sumar(5, 3) = {c.sumar(5, 3)}")
    print(f"sumar(5, 3, 2) = {c.sumar(5, 3, 2)}")

    print(f"\nmultiplicar(4, 5) = {c.multiplicar(4, 5)}")
    print(f"multiplicar(4, 5, 2) = {c.multiplicar(4, 5, 2)}")

    print(f"\noperar(1, 2, 3, 4, 5) = {c.operar(1, 2, 3, 4, 5)}")

File: 1erdepartamental/c++/4.sobrecarga.cpp
// ============================================
// PROGRAMA: Sobrecarga con POO
// DESCRIPCIÓN: Simula sobrecarga usando parámetros por defecto
// ============================================

#include <iostream>
#include <stdexcept>
using namespace std;


/**
 * Interfaz para operadores flexibles
 */
class IOperadorFlexible {
public:
    virtual double operar() = 0;
    virtual ~IOperadorFlexible() {}
};


/**
 * Simula sobrecarga con parámetros por defecto
 * 
 * C++ SÍ soporta sobrecarga nativa, pero se demuestra con defaults
 */
class CalculadoraSobrecarga : public IOperadorFlexible {
private:
    double __resultado;

public:
    CalculadoraSobrecarga() : __resultado(0) {}
    
    double getResultado() const { return __resultado; }
    
    // "Sobrecarga" con parámetros por defecto
    double sumar(double a = 0, double b = 0, double c = 0) {
        __resultado = a + b + c;
        return __resultado;
    }
    
    double restar(double a = 0, double b = 0, double c = 0) {
        __resultado = a - b - c;
        return __resultado;
    }
    
    double multiplicar(double a = 1, double b = 1, double c = 1) {
        __resultado = a * b * c;
        return __resultado;
    }
    
    double dividir(double a = 1, double b = 1, double c = 1) {
        if (b != 0 && c != 0) {
            __resultado = a / b / c;
            return __resultado;
        }
        throw runtime_error("División por cero");
    }
    
    double operar() override {
        return __resultado;
    }
    
    friend ostream& operator<<(ostream& os, const CalculadoraSobrecarga& c) {
        os << "CalculadoraSobrecarga(resultado=" << c.__resultado << ")";
        return os;
    }
};


// ============================================
// EJECUCIÓN
// ============================================
int main() {
    CalculadoraSobrecarga c;
    
    // Demuestra "sobrecarga" con diferente cantidad de argumentos
    cout << "sumar() = " << c.sumar() << endl;
    cout << "sumar(5, 3) = " << c.sumar(5, 3) << endl;
    cout << "sumar(5, 3, 2) = " << c.sumar(5, 3, 2) << endl << endl;
    
    cout << "multiplicar(4, 5) = " << c.multiplicar(4, 5) << endl;
    cout << "multiplicar(4, 5, 2) = " << c.multiplicar(4, 5, 2) << endl << endl;
    
    cout << c << endl;
    
    return 0;
}
